[JENKINS-68649] – Azure ACI Connector compatibility changes

hWorblehat opened the pull request:
This PR relaxes the requirement for Pods to be `Running` while the
agent launcher is waiting for the agent to connect. This should address
[JENKINS-68649], which is caused by pods on virtual nodes still being in
the `Pending` for a short while once all of its containers are running.

Fixes: JENKINS-68649

- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [X] **(tested manually)** Ensure you have provided tests - that demonstrates feature works or fixes the issue

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md  in your own repository 
-->

[JENKINS-68649]: https://issues.jenkins.io/browse/JENKINS-68649

Vlatombe commented:
This is incorrect. The agent should not be usable until it's running and all containers are up.

hWorblehat commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:
```diff
@@ -176,7 +176,7 @@ else if (httpCode == 409 && e.getMessage().contains("Operation cannot be fulfill
 
             // We need the pod to be running and connected before returning
             // otherwise this method keeps being called multiple times
```
I'm assuming your assessment that my change is incorrect is based on this comment. This threw me at first as well, but if you look at how the `validStates` list is actually used below:

 - it's **not** used for _waiting_ on the pod at all
 - it's used for checking that the pod has entered a "broken" state - if the pod enters a state not on this list, the code _stops_ waiting, the pod is assumed broken, and is destroyed immediately. This is what's causing the bug I'm seeing.

My best guess was that the comment refers to the entire wait loop, not just the `validStates` list that it's immediately above.

hWorblehat reviewed (commented):
Hi @Vlatombe,

> The agent should not be usable until it's running and all containers are up.

I'm not disputing that, and I'm not intending this PR to dispute that either!

hWorblehat commented:
The error that the plugin code is currently making seems to be assuming that having all containers `Running` implies that the pod must also have the `Running` state/phase. This does not appear to be true for pods on virtual nodes (see the linked bug).

If you prefer, I could change the PR so that the wait loop _does_ wait on the running state instead, but that's not what the plugin currently does.

hWorblehat commented:
I've updated the PR to check for terminated statuses as discussed. I've also changed the way the wait is terminated in that case, to avoid incorrectly counting metrics:

Previously, we were ending up in [this code block](https://github.com/jenkinsci/kubernetes-plugin/pull/1188/files#diff-8e43be7dacff0fa974a7614fda2f4103bf4b820be52d747d9498e6f7efbd688fR242-R249), which would have cause the `LAUNCH_FAILED` metric to be double-counted, and the `FAILED_TIMEOUT` metric to be incremented incorrectly. Now, instead of `break`ing out fo the loop, I throw an exception directly.

Having done all of that, I'm not sure this pod status check is necessary at all. We check whether each of the _containers_ are terminated [just below](https://github.com/jenkinsci/kubernetes-plugin/pull/1188/files#diff-8e43be7dacff0fa974a7614fda2f4103bf4b820be52d747d9498e6f7efbd688fR215-R231), so we could remove the pod status check entirely, and rely on the container status checks instead.

hWorblehat commented:
Apologies - links to code block did not work correctly in previous comment. Have now edited to just refer to line numbers...

hWorblehat commented:
Hi @Vlatombe ,

Can you confirm if the CI failure is genuine? I've seen that same test failing in with the same error in a couple of other PRs (including a documentation update), which makes me wonder if it's just a flaky test?

> Expected: a string containing "jnlp -- terminated (1)"
>     but: was "Started

Vlatombe commented:
Yes, it's a flake.
